src/lib/shim/src/shimlogger.rs

```diff
@@ -0,0 +1,148 @@
+                BorrowedFd::borrow_raw(linux_raw_sys::general::STDOUT_FILENO.try_into().unwrap())
+            }),
+        }
+    }
+
+    pub fn flush(&mut self) {
+        self.stdout.write_str(self.buffer.as_str()).unwrap();
+        self.buffer.reset();
+    }
+}
+
+impl core::fmt::Write for ShimLoggerWriter {
+    fn write_str(&mut self, s: &str) -> Result<(), core::fmt::Error> {
+        if s.len() > self.buffer.capacity_remaining() {
+            // Flush to make room and keep output FIFO.
+            self.flush();
+        }
+        if s.len() > self.buffer.capacity_remaining() {
+            // There will never be enough room. Write directly to stdout.
+            self.stdout.write_str(s)
+        } else {
+            // Write to buffer. Should be impossible to fail.
+            self.buffer.write_str(s)
+        }
+    }
+}
+
+impl Drop for ShimLoggerWriter {
+    fn drop(&mut self) {
+        self.flush();
+    }
+}
+
+/// Implementation of `log::Log` for use in the shim. Includes some shim related
+/// metadata (such as simulation time), is no_std, and is careful to only make
+/// inlined syscalls to avoid getting intercepted by the shim's seccomp filter.
+pub struct ShimLogger {}
+
+impl ShimLogger {
+    /// Install a `ShimLogger` as  the logging backend in the Rust `log` crate.
+    /// Should only be called once.
+    pub fn install(log_level: log::LevelFilter) {
+        // log::set_logger requires a logger with a static lifetime.
+        const SHIM_LOGGER: ShimLogger = ShimLogger {};
+
+        log::set_max_level(log_level);
+        log::set_logger(&SHIM_LOGGER).unwrap();
+    }
+}
+
+impl log::Log for ShimLogger {
+    fn enabled(&self, metadata: &log::Metadata) -> bool {
+        metadata.level() <= log::max_level()
+    }
+
+    fn log(&self, record: &log::Record) {
+        if !self.enabled(record.metadata()) {
+            return;
+        }
+
+        let mut writer = ShimLoggerWriter::new();
+
+        match crate::manager_shmem() {
+            Some(m) => {
+                // rustix's `clock_gettime` goes through VDSO, which is overwritten with our trampoline,
+                // which would end up trying to log and recurse.
+                // `linux_api`'s `clock_gettime` always makes the syscall, which is what we want.
+                let now = linux_api::time::clock_gettime(linux_api::time::ClockId::CLOCK_REALTIME)
+                    .unwrap();
+
+                let now = Duration::from_secs(now.tv_sec.try_into().unwrap())
+                    + Duration::from_nanos(now.tv_nsec.try_into().unwrap());
+
+                let start = Duration::from_micros(m.log_start_time_micros.try_into().unwrap());
+                let elapsed = now - start;
+                let parts = TimeParts::from_nanos(elapsed.as_nanos());
+                core::fmt::write(
+                    &mut writer,
+                    format_args!("{} ", parts.fmt_hr_min_sec_nano()),
+                )
+                .unwrap();
```

Comment:
nit: just `write!(&mut writer, "{} ", parts.fmt_hr_min_sec_nano())`? (and for other writes below)

Reply:
ah, yeah that's nicer